viral/rastermap_utils.py

```diff
@@ -0,0 +1,488 @@
+        trial (TrialInfo):              A TrialInfo object representing the trial to be analysed.
+        trial_frames (np.ndarray):      Numpy array representing the frames in the trial.
+        wheel_circumference (float):    The circumference of the wheel on the rig used in centimetres.
+        ITI (bool, optional):           Include the frames in the inter-trial interval (ITI). Defaults to False.
+
+    Returns:
+        np.ndarray:                     A Numpy array of shape (frames, 2). Each row representing a frame with:
+            frame_idx (int)
+            position (float)
+    """
+    if ITI:
+        frames_start_end = np.array(
+            [
+                [state.closest_frame_start, state.closest_frame_end]
+                for state in trial.states_info
+                if state.name
+                in ["trigger_panda", "trigger_panda_post_reward", "trigger_panda_ITI"]
+            ]
+        )
+    else:
+        frames_start_end = np.array(
+            [
+                [state.closest_frame_start, state.closest_frame_end]
+                for state in trial.states_info
+                if state.name in ["trigger_panda", "trigger_panda_post_reward"]
+            ]
+        )
+    degrees = np.array(trial.rotary_encoder_position)
+    positions = degrees_to_cm(degrees, wheel_circumference)
+    assert len(frames_start_end) == len(positions)
+    frames_positions_combined = list()
+    for i in range(len(frames_start_end)):
+        frames_positions_combined.append([frames_start_end[i, 0], positions[i]])
+        frames_positions_combined.append([frames_start_end[i, 1], positions[i]])
+    frames_positions_combined = np.array(frames_positions_combined)
+    unique_frames_positions = np.unique(frames_positions_combined, axis=0)
+    # Interpolation is needed as the position is not stored when the reward comes on, hence position for several frames might be missing
+    interp_func = interp1d(
+        unique_frames_positions[:, 0],
+        unique_frames_positions[:, 1],
+        kind="linear",
+        fill_value="extrapolate",
+    )
+    interpolated_positions = interp_func(trial_frames)
+    assert len(trial_frames) == len(
+        interpolated_positions
+    ), "# of trial frames is unequal to # of interpolated positions"
+    return np.column_stack((trial_frames, interpolated_positions))
+
+
+def get_speed_frame(frame_position: np.ndarray) -> np.ndarray:
+    # TODO: Is this ok like this?
+    frames_diff = np.diff(frame_position[:, 0])
+    positions_diff = np.diff(frame_position[:, 1])
+    speeds = positions_diff / frames_diff
+    # little hack, have to account for one more frame
+    last_speed = speeds[-1] if len(speeds) > 0 else 0
+    # Add last frame with speed of the last computed frame
+    new_frame = np.array([[frame_position[-1, 0], last_speed]])
+    return np.vstack([np.column_stack((frame_position[:-1, 0], speeds)), new_frame])
+
+
+def get_lick_index(trial: TrialInfo) -> np.ndarray | None:
```

Comment:
(I know this is hypocritical). Could we get a little comment about what the function does? From the test, it looks like it returns frames on which a lick occured. Is that right?